pkg/apis/pubsub/v1alpha1/topic_lifecycle.go

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/apis/duck/v1alpha1"
+)
+
+// GetCondition returns the condition currently associated with the given type,
+// or nil.
+func (ts *TopicStatus) GetCondition(t apis.ConditionType) *apis.Condition {
+	return topicCondSet.Manage(ts).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (ts *TopicStatus) IsReady() bool {
+	return topicCondSet.Manage(ts).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (ts *TopicStatus) InitializeConditions() {
+	topicCondSet.Manage(ts).InitializeConditions()
+}
+
+// TODO: Use the new beta duck types.
+func (ts *TopicStatus) SetAddress(url *apis.URL) {
+	if ts.Address == nil {
+		ts.Address = &v1alpha1.Addressable{}
+	}
+	if url != nil {
+		ts.Address.Hostname = url.Host
+		ts.Address.URL = url
+		topicCondSet.Manage(ts).MarkTrue(TopicConditionAddressable)
+	} else {
+		ts.Address.Hostname = ""
+		ts.Address.URL = nil
+		topicCondSet.Manage(ts).MarkFalse(TopicConditionAddressable, "emptyHostname", "hostname is the empty string")
+	}
+}
+
+func (ts *TopicStatus) PropagateDeploymentAvailability(d *appsv1.Deployment) {
+	for _, cond := range d.Status.Conditions {
+		if cond.Type == appsv1.DeploymentAvailable {
+			switch cond.Status {
+			case "True":
+				ts.MarkDeployed()
+			case "False":
+				ts.MarkNotDeployed(cond.Reason, cond.Message)
+			default:
+				ts.MarkDeploying(cond.Reason, cond.Message)
+			}
+			return
+		}
+	}
+	ts.MarkDeploying("DeploymentStatus", "Failed to inspect Deployment status.")
```

Comment:
This message should be clearer about what actually happened (the deployment didn't have the expected condition).

Reply:
Sure, and that should be explained in the message. :)